### ISSUE TITLE
[Backport] rancher-webhook k8s 1.28 chart support

### ISSUE THREAD

**Issue by MKlimuszka** [rancher/rancher#44018]
**This issue is to implement the chart changes & perform the required testing to support k8s 1.28. The generic process that's applicable for most charts is described here. Please adjust the steps to fit your respective chart wherever required. Perform the steps below:**

- [x] Once you pick up the issue for your chart, apply the label `[zube]: Working` to the issue
Note: If this is an upstream chart (ex: istio) and if the upstream chart doesn't support k8s 1.28 then this exercise is not applicable to you. Just add a comment that the upstream chart doesn't support k8s 1.28 and apply the label `[zube]: Done`. If this is not an upstream chart then proceed to the next step.
- [ ] Update kube-version annotation in Chart.yaml to `catalog.cattle.io/kube-version: '>= 1.23.0-0'` or `catalog.cattle.io/kube-version: '>= 1.23.0-0 < 1.29.0-0'`
- [ ] Follow the usual process to add a new chart version and file a PR to the `rancher/charts:dev-v2.8` branch. Link the PR to the issue.
- [ ] Perform the usual Developer testing and if everything looks good then merge the PR. After merging the PR,  apply the label `[zube]: To Test`. Tag the respective QA team members.
- [ ] Perform below testing
   - [ ] Spin up Rancher 2.8 with k8s 1.27
   - [ ] Add a downstream cluster with k8s 1.28
   - [ ] Point the Rancher charts repository of the downstream cluster to dev-v2.8 branch
   - [ ] Install the dev version of the chart. Verify that the installation is successful
   - [ ] Perform chart functionality testing (any tests in QA test plans for the chart)
   - [ ] Uninstall the chart. Verify that the uninstallation is successful
   - [ ] Install the prod version of the chart. Verify that the installation is successful
   - [ ] Upgrade to the dev version of the chart. Verify that the upgrade is successful
   - [ ] Add a downstream cluster with k8s 1.27
   - [ ] Install the prod version of the chart
   - [ ] Upgrade the k8s version of the downstream cluster to 1.28. Verify the chart still works after the upgrade
   - [ ] Perform any additional testing specific to your chart if required
- [ ] If the testing passes then apply the label `[zube]: Done`. If you are blocked with any steps, add a comment with details and apply the label `[zube]: Blocked`

SURE-7716

**Comment by Priyashetty17** [comment]
@samjustus Isn't this a duplicate of https://github.com/rancher/rancher/issues/43728?

**Comment by tomleb** [comment]
We have decided to complete https://github.com/rancher/rancher/issues/44619 first in order to add support for k8s 1.28. 

**Here is why:**


cluster-api `v1.5.5` calls `validation.NewSchemaValidator`  [somewhere in its webhook validation](https://github.com/kubernetes-sigs/cluster-api/blob/84b6eee92da051c9ea7ed0c4c0e6da9c56f8fa4f/internal/webhooks/cluster.go#L492).This function signature (defined in apiextensions repo) changed between [1.27](https://github.com/kubernetes/apiextensions-apiserver/blob/v0.27.4/pkg/apiserver/validation/validation.go#L33) and [1.28](https://github.com/kubernetes/apiextensions-apiserver/blob/v0.28.6/pkg/apiserver/validation/validation.go#L61):

```go
// 1.27
func NewSchemaValidator(customResourceValidation *apiextensions.CustomResourceValidation) (*validate.SchemaValidator, *spec.Schema, error) { /*...*/ }
// 1.28
func NewSchemaValidator(customResourceValidation *apiextensions.JSONSchemaProps) (SchemaValidator, *spec.Schema, error)  { /*...*/ }
```

Since [cluster-api v1.5.5 imports k8s 1.27](https://github.com/kubernetes-sigs/cluster-api/blob/v1.5.5/go.mod#L37), it's using the first function signature. But since rancher/webhook is importing v0.28.6, go tries to use the second function signature, and fails because cluster-api is passing a `*CustomResourceValidation` instead of a `*JSONSchemaProps`. rancher/rancher builds fine because it doesn't import [sigs.k8s.io/cluster-api/webhooks](http://sigs.k8s.io/cluster-api/webhooks), but I tested importing that in rancher/rancher and building and I get the same error.

To solve this, I have tried the following:

- **Bumping cluster-api to v1.6.X**

This forces us to use controller-runtime v0.16.3 instead of v0.15.3 which would require some simple changes in lasso but also more changes in webhook itself. Looks like a few fields changed/were removed. I didn't go all the way to get something to compile. Seems like a lot of work. And as @snasovich points out, this would have a ripple effect in rancher/rancher.

- **Forcing apiextensions to 1.27**

This also requires forcing [k8s.io/apiserver](http://k8s.io/apiserver) to 1.27. We end up with this frankenstein go.mod that is both 1.28 and 1.27. Unsure if we can consider this "supporting 1.28". It compiles, haven't run the unit/integration tests to see if anything weird comes up.

None of these are great.

**Here's what we agreed to do:**

We're going to be [removing CAPI webhook ](https://github.com/rancher/rancher/issues/44619) out of webhook. Go to that issue for more details. This will allow us to NOT import the webhook specific path that causes the conflict.

**Comment by tomleb** [comment]
@MKlimuszka Still in review, needs to be bumped in  rancher/rancher fyi

**Comment by Priyashetty17** [comment]
Validated with `v2.8-0b0d677fa3c980137dcc1cd120b203d154728d21-head`

<a name="top"></a>

### :muscle: Test Cases

| \#  | Priority | Description & Link                           | PASS/FAIL |
| --- | -------- | -------------------------------------------- | --------- |
| 1   | P0       | [Fresh Install Rancher - k8s v1.28](#test-1) | ✅ PASS    |
| 2   | P0       | [Upgrade - k8s 1.27 to 1.28](#test-2)        | ✅ PASS    |

<details>
    <summary>🚨 2 test cases... CLICK TO EXPAND! (For table links to work) ⬅️</summary>

<br>

# 1 / Fresh Install Rancher - k8s v1.28 Status: ✅ PASS

<a name="test-1"></a>

**:small_red_triangle: [back to top](#top)**

<details>
    <summary>Test 1 details... Click to expand</summary>

**Test Steps for Validation**

1. Create a rancher server v2.8-head with k8s v1.28 and a downstream cluster with k8s 1.28.
2. Verify that rancher-webhook is installed on both local and downstream clusters
3. Verify that the version of the webhook is as expected on both local and downstream clusters- `v0.4.3-rc1`
4. Verify kube-version annotation in the webhook chart's yaml has: catalog.cattle.io/kube-version: '>= 1.23.0-0 < 1.29.0-0'
5. Verify that there are no errors in the rancher-webhook logs
6. Verify count in local cluster with `kubectl get validatingwebhookconfigurations.admissionregistration.k8s.io rancher.cattle.io`
   1. response: `rancher.cattle.io   16`
   2. passing expected count
7. Verify count in downstream cluster with `kubectl get validatingwebhookconfigurations.admissionregistration.k8s.io rancher.cattle.io`
   1. response: `rancher.cattle.io   7`
   2. passing expected count
8. Create a restricted admin user
9. Login as restricted admin
10. Attempt to edit the admin role, while logged in as restricted admin
11. Attempt to set admin to "Yes: Default role for new users"
    1.  Blocked by webhook as expected

**✅ Expected Outcome**

For Rancher webhook to work with k8s 1.28. For the expected version to be installed and for there to be no errors present. For webhook P0 checks to pass.

**✅ Actual Outcome**

Webhook checks are passing as expected.

</details>

<hr>

# 2 / Upgrade - k8s 1.27 to 1.28 Status: ✅ PASS

<a name="test-2"></a>

**:small_red_triangle: [back to top](#top)**

<details>
    <summary>Test 2 details... Click to expand</summary>

**Test Steps for Validation**

1. Create a rancher server on the latest released version with k8s v1.27 
2. Create a downstream RKE2 cluster with k8s 1.27
3. Upgrade Rancher to v2.8-head
4. Upgrade the downstream RKE2 cluster from 1.27 to ==> 1.28
5. Verify that rancher-webhook is upgraded on both local and downstream clusters
6. Verify that the version of the webhook is as expected on both local and downstream clusters. `v0.4.3-rc1`
7. Verify kube-version annotation in the webhook chart's yaml has: catalog.cattle.io/kube-version: '>= 1.23.0-0 < 1.29.0-0'
8. Verify that there are no errors in the rancher-webhook logs
9. Verify count in local cluster with `kubectl get validatingwebhookconfigurations.admissionregistration.k8s.io rancher.cattle.io`
   1. response: `rancher.cattle.io   16`
   2. passing expected count
10. Verify count in downstream cluster with `kubectl get validatingwebhookconfigurations.admissionregistration.k8s.io rancher.cattle.io`
   1. response: `rancher.cattle.io   7`
   2. passing expected count
11. Create a restricted admin user
12. Login as restricted admin
13. Attempt to edit the admin role, while logged in as restricted admin
14. Attempt to set admin to "Yes: Default role for new users"
    1.  Blocked by webhook as expected

**✅ Expected Outcome**

For Rancher webhook to work with k8s 1.28 after upgrade. For the expected version to be installed and for there to be no errors present. For webhook P0 checks to pass.

**✅ Actual Outcome**

Webhook checks are passing as expected.


</details>

<hr>

</details>
